### PR TITLE
Migrate from PrestoSQL to Trino 

### PR DESCRIPTION
**Pull request by akshayrai**
Issue ref: #66

Couple of major call-outs from the PrestoSql (333) -> Trino (352) upgrade
* The  SqlScalarFunction in the non-spi package has undergone interface change
The SqlScalarFunction interface no longer exposes the global Trino metadata but exposes only the relevant FunctionDependencies. ([upstream trino pr reference](https://github.com/trinodb/trino/pull/3970/files)). The code has been modified accordingly by overriding the `getFunctionDependencies` method and declaring the input/output function parameters and operators. The test framework has been fixed by allowing the metadata to still be supported in the TrinoFactory.

* Rename from presto/sql to trino
This has been done is a separate commit. Please pick relevant commit history and review accordingly.

* Trino has a build requirement on jdk 11 (addressed in a separate PR: https://github.com/linkedin/transport/pull/69)
Since Spark and Hive still need to compile with jdk8, we need to isolate and build the artifacts with different jdks. This is achieved by leveraging gradle toolchains and building the compile and test tasks with respective jdks.

### CODE REVIEW

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java [hunk 1]
```diff
@@ -110,14 +144,17 @@ public ScalarFunctionImplementation specialize(BoundVariables boundVariables, in
         - (new Random()).nextInt(initialJitterInt));
     boolean[] nullableArguments = stdUDF.getAndCheckNullableArguments();
 
-    return new ScalarFunctionImplementation(true, getNullConventionForArguments(nullableArguments),
-        getMethodHandle(stdUDF, metadata, boundVariables, nullableArguments, requiredFilesNextRefreshTime));
+    return new ChoicesScalarFunctionImplementation(
+        functionBinding,
+        FAIL_ON_NULL,
```
Comment (shardulm94) [inline transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java:149]: If arguments of a Transport UDF are defined as non-nullable and if a null value is passed to it, the current behaviour is that the UDF will return a null. In Trino, previously we didn't have any special handling on Transport side for this as Trino behaviour was also the same. Does this change that behaviour to fail instead if nulls are passed?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java [hunk 2]
```diff
@@ -98,8 +106,34 @@ protected long getRefreshIntervalMillis() {
...
+  public FunctionDependencyDeclaration getFunctionDependencies(FunctionBinding functionBinding) {
+    FunctionDependencyDeclaration.FunctionDependencyDeclarationBuilder builder = FunctionDependencyDeclaration.builder();
+
+    List<Type> types = new ArrayList<>();
+    types.add(functionBinding.getBoundSignature().getReturnType());
+    types.addAll(functionBinding.getBoundSignature().getArgumentTypes());
+
+    types.forEach(type -> builder.addType(type.getTypeSignature()));
+    if (containsMapType(types)) {
+      builder.addOperatorSignature(EQUAL, ImmutableList.of(new TypeSignature("K"), new TypeSignature("K")));
```
Comment (shardulm94) [inline transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java:118]: Is the type signature being hardcoded to `K` correct? Shouldn't it apply to the type of the map key? What if there are multiple map types with different map key types?

File: docs/transport-udfs-trino.patch [hunk 3]
```diff
@@ -1,23 +1,23 @@
-diff --git a/presto-main/src/main/java/io/prestosql/server/PluginManager.java b/presto-main/src/main/java/io/prestosql/server/PluginManager.java
+diff --git a/trino-main/src/main/java/io/trino/server/PluginManager.java b/trino-main/src/main/java/io/trino/server/PluginManager.java
```
Comment (shardulm94) [inline docs/transport-udfs-trino.patch:1]: I tried to apply this patch to Trino release tag 352 and the apply failed.

```
$ ~/trino @ 690ddcd090 > git status
HEAD detached at 352
nothing to commit, working tree clean
$ ~/trino @ 690ddcd090 > git apply ~/transport/docs/transport-udfs-trino.patch
error: corrupt patch at line 48
```
Can you take a look?

File: README.md [hunk 4]
```diff
@@ -4,13 +4,13 @@
 **Transport** is a framework for writing performant user-defined
 functions (UDFs) that are portable across a variety of engines
 including [Apache Spark](https://spark.apache.org/), [Apache Hive](https://hive.apache.org/), and
-[Presto](https://prestodb.io/). Transport UDFs are also
+[Trino](https://trinodb.io/). Transport UDFs are also
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply README.md]: fixed

File: defaultEnvironment.gradle [hunk 5]
```diff
@@ -10,7 +10,7 @@ subprojects {
       url "https://conjars.org/repo"
     }
   }
-  project.ext.setProperty('presto-version', '333')
+  project.ext.setProperty('trino-version', '352')
   project.ext.setProperty('airlift-slice-version', '0.38')
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply defaultEnvironment.gradle]: updated to 0.39

File: docs/using-transport-udfs.md [hunk 6]
```diff
@@ -80,16 +80,16 @@ If the UDF class is `com.linkedin.transport.example.ExampleUDF` then the platfor
...
 
-### Presto
+### Trino
 
-1. Add the UDF to the Presto installation  
-Unlike Hive and Spark, Presto currently does not allow dynamically loading jar files once the Presto server has started.
-In Presto, the jar is deployed to the `plugin` directory.
-However, a small patch is required for the Presto engine to recognize the jar as a plugin, since the generated Presto UDFs implement the `SqlScalarFunction` API, which is currently not part of Presto's SPI architecture.
-You can find the patch [here](transport-udfs-presto.patch) and apply it before deploying your UDFs jar to the Presto engine.
+1. Add the UDF to the Trino installation
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply docs/using-transport-udfs.md]: fixed

File: transportable-udfs-plugin/src/main/java/com/linkedin/transport/plugin/packaging/ThinJarPackaging.java [hunk 7]
```diff
@@ -25,7 +25,7 @@
   public List<TaskProvider<? extends Task>> configurePackagingTasks(Project project, Platform platform,
       SourceSet platformSourceSet, SourceSet mainSourceSet) {
     /*
-      task <platformName>ThinJar(type: Jar, dependsOn: prestoClasses) {
+      task <platformName>ThinJar(type: Jar, dependsOn: trinoClasses) {
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply transportable-udfs-plugin/src/main/java/com/linkedin/transport/plugin/packaging/ThinJarPackaging.java]: updated

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 1 above)
Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: updated it to NULLABLE_RETURN.

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 2 above)
Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: good point! updated it to the map key type

File: docs/transport-udfs-trino.patch (hunk 3 above)
Comment (akshayrai) [reply docs/transport-udfs-trino.patch]: thanks for verifying! fixed the patch

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 1 above)
Comment (shardulm94) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: I am surprised there wasn't a case in the `examples` project to catch this. Would it be possible to add a test for this?

I think its fine if it is not addressed as part of this PR, but atleast we should make sure we aren't causing a regression here.

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+import static io.trino.sql.analyzer.TypeSignatureTranslator.*;
+
+
+public class TrinoFactory implements StdFactory {
+
+  final FunctionBinding functionBinding;
+  final FunctionDependencies functionDependencies;
+  final Metadata metadata;
+
+  public TrinoFactory(FunctionBinding functionBinding, FunctionDependencies functionDependencies) {
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java:58]: Ok, I am not sure where I saw `getMetadata`. I also see that `FunctionDependencies` takes `Metadata` object in the constructor. Can we use that instead within `TrinoTester` instead to get rid of the `Metadata`-based constructor?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 2 above)
Comment (shardulm94) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: I am not sure if this is fully correct. For nested types we may need to recurse through the type signatures to identify map keys within struct types. Is there are a good way we can test this?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java:58]: Explored that route - but the getType operation which transport depends on, within FunctionDependencies, is derived from the typeDependencies (2nd argument) and not from the metadata. So encapsulating just the metadata within FunctionDependencies is not enough. On the other hand, TrinoTester is unaware of the typeDependencies (function arguments & return types) and hence unable to register them.

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 1 above)
Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: > If arguments of a Transport UDF are defined as non-nullable and if a null value is passed to it, the current behaviour is that the UDF will return a null. 

We already [have tests ](https://github.com/linkedin/transport/blob/master/transportable-udfs-examples/transportable-udfs-example-udfs/src/test/java/com/linkedin/transport/examples/TestArrayElementAtFunction.java#L33)to validate the above behaviour but it looks like this field doesn't govern that behaviour.

This parameter (NULLABLE_RETURN) and the subsequent one `getNullConventionForArguments` only define the convention and used to validate if the function can be adapted to this pattern.

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 2 above)
Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: Can you share an example of this?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (shardulm94) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java:58]: Thanks for looking into this

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java (hunk 2 above)
Comment (shardulm94) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: e.g. a type like with `struct<field:map<K,V>>`

Comment (akshayrai) [reply transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java]: registered nested types as well.

**Review by shardulm94 — APPROVED** [review]

**Comment by CHENXCHEN** [comment]
Does it support both PrestoSQL and Trino ?
Some version in PrestoSQL used in production...

**Pull request merged**